pennylane/devices/qutrit_mixed/sampling.py

```diff
@@ -0,0 +1,334 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for sampling a qutrit mixed state.
+"""
+from typing import List, Union
+
+import numpy as np
+import pennylane as qml
+from pennylane import math
+from pennylane.measurements import (
+    Shots,
+    SampleMeasurement,
+    SampleMP,
+    CountsMP,
+)
+from pennylane.typing import TensorLike
+
+from .utils import QUDIT_DIM, get_num_wires
+from .measure import measure
+from .apply_operation import apply_operation
+
+
+# pylint:disable = too-many-arguments
+def _sample_state_jax(
+    state,
+    shots: int,
+    prng_key,
+    is_state_batched: bool = False,
+    wires=None,
+) -> np.ndarray:
+    """Returns a series of samples of a state for the JAX interface based on the PRNG.
+
+    Args:
+        state (array[complex]): A state vector to be sampled
+        shots (int): The number of samples to take
+        prng_key (jax.random.PRNGKey): A``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator.
+        is_state_batched (bool): whether the state is batched or not
+        wires (Sequence[int]): The wires to sample
+
+    Returns:
+        ndarray[int]: Sample values of the shape (shots, num_wires)
+    """
+    # pylint: disable=import-outside-toplevel
+    import jax
+    import jax.numpy as jnp
+
+    key = prng_key
+
+    total_indices = get_num_wires(state, is_state_batched)
+    state_wires = qml.wires.Wires(range(total_indices))
+
+    wires_to_sample = wires or state_wires
+    num_wires = len(wires_to_sample)
+    basis_states = np.arange(QUDIT_DIM**num_wires)
+
+    with qml.queuing.QueuingManager.stop_recording():
+        probs = measure(qml.probs(wires=wires_to_sample), state, is_state_batched)
+
+    if is_state_batched:
+        # Produce separate keys for each of the probabilities along the broadcasted axis
+        keys = []
+        for _ in state:
+            key, subkey = jax.random.split(key)
+            keys.append(subkey)
+        samples = jnp.array(
+            [
+                jax.random.choice(_key, basis_states, shape=(shots,), p=prob)
+                for _key, prob in zip(keys, probs)
+            ]
+        )
+    else:
+        samples = jax.random.choice(key, basis_states, shape=(shots,), p=probs)
+
+    res = np.zeros(samples.shape + (num_wires,), dtype=np.int64)
```

Comment:
From what I can tell, this is the same as the one in the qubit device except for line 67 (with `QUDIT_DIM`) and these few lines here, right?

Out of curiosity, why does JAX require a separate function?

Reply:
Jax requires a separate function because it has it's own random number generator, it is necessary for jitting and jax grads.